### PR TITLE
NRF51 UART send_bytes working

### PR DESCRIPTION
**Pull request by amilich**

### CODE REVIEW

File: boards/nrf51dk/src/main.rs [hunk 1]
```diff
@@ -159,6 +162,19 @@ pub unsafe fn reset_handler() {
         24);
     nrf51::uart::UART0.set_client(console);
 
+    nrf51::uart::UART0.init(kernel::hil::uart::UARTParams {
+        baud_rate: 115200,
+        data_bits: 8,
+        parity: kernel::hil::uart::Parity::None,
+        mode: kernel::hil::uart::Mode::Normal,
+    });
+    nrf51::uart::UART0.enable_tx();
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: I don't think this line is necessary/useful

File: boards/nrf51dk/src/main.rs [hunk 2]
```diff
@@ -205,6 +221,9 @@ pub unsafe fn reset_handler() {
     let mut chip = nrf51::chip::NRF51::new();
     chip.systick().reset();
     chip.systick().enable(true);
+
+    nrf51::uart::UART0.send_bytes(&mut bytes, bytes.len());
+
```
Comment (alevy) [inline boards/nrf51dk/src/main.rs]: Please remove test code

File: chips/nrf51/src/uart.rs [hunk 3]
```diff
@@ -142,7 +159,16 @@ impl UART {
             }
         }
         if tx {
-            // Should never execute!
+            if self.len.get() == self.index.get() {
+                regs.task_stoptx.set(1 as u32);
```
Comment (alevy) [inline chips/nrf51/src/uart.rs]: You need to deliver this event up to the `client`'s `send_done` method

**Comment by amilich** [comment]
Just made those changes, let me know if there's anything else I should do. I think printf still wasn't working even when client.write_done is called.

File: chips/nrf51/src/uart.rs [hunk 4]
```diff
@@ -142,7 +159,25 @@ impl UART {
...
         if tx {
-            // Should never execute!
+            regs.event_txdrdy.set(0 as u32);
+
+            if self.len.get() == self.index.get() {
+                regs.task_stoptx.set(1 as u32);
+
+                // Signal client write done
+                match self.client {
+                    Some(ref client) => client.write_done(self.buffer.take().unwrap()),
```
Comment (alevy) [inline chips/nrf51/src/uart.rs]: No `unwrap`. Use `map` or `unwrap_or` instead. `unwrap` can panic and we should avoid potential panics whenever possible.

File: chips/nrf51/src/uart.rs [hunk 5]
```diff
@@ -142,7 +159,25 @@ impl UART {
...
         }
         if tx {
-            // Should never execute!
+            regs.event_txdrdy.set(0 as u32);
+
+            if self.len.get() == self.index.get() {
+                regs.task_stoptx.set(1 as u32);
+
+                // Signal client write done
+                match self.client {
```
Comment (alevy) [inline chips/nrf51/src/uart.rs]:
```rust
self.client.map(|client| {
   ...
});
```

Would accomplish this more cleanly than a `match` with an empty `None` condition (it's actually exactly the same thing)

**Comment by amilich** [comment]
Sorry - just saw this last one. I pushed a fix with map (I had to use buffer.take())

**Pull request merged**